warehouse/query-core/src/main/java/datawave/next/DocIdQueryIterator.java

```diff
@@ -0,0 +1,324 @@
+
+    public static final String BATCH_SIZE = "batch.size";
+    public static final String SCAN_TIMEOUT = "scan.timeout";
+
+    private Range range;
+    private ASTJexlScript script;
+    private Set<String> datatypeFilter;
+    private LongRange timeFilter;
+
+    private Set<String> indexedFields;
+
+    private SortedKeyValueIterator<Key,Value> source;
+    private Map<String,String> options;
+    private IteratorEnvironment env;
+    private int batchSize = 1;
+    private long scanTimeout = -1;
+
+    private Key tk;
+    private Value tv = new Value();
+
+    private Iterator<Key> data;
+
+    private boolean statsReturned = false;
+    private final DocIdQueryIteratorStats timingStats = new DocIdQueryIteratorStats();
+    private final DocumentIteratorStats iteratorStats = new DocumentIteratorStats();
+
+    public DocIdQueryIterator() {}
+
+    public DocIdQueryIterator(DocIdQueryIterator other, IteratorEnvironment env) {
+        this.source = other.source.deepCopy(env);
+        this.options = other.options;
+        this.env = other.env;
+
+        this.range = other.range;
+        this.script = other.script;
+        this.datatypeFilter = other.datatypeFilter;
+        this.timeFilter = other.timeFilter;
+    }
+
+    @Override
+    public void init(SortedKeyValueIterator<Key,Value> source, Map<String,String> options, IteratorEnvironment env) throws IOException {
+        timingStats.markScanInit();
+        this.source = source;
+        this.options = options;
+        this.env = env;
+
+        // need the query, datatype filter, start and end dates (for time filter)
+
+        if (options.containsKey(QueryOptions.QUERY)) {
+            String opt = options.get(QueryOptions.QUERY);
+
+            try {
+                this.script = JexlASTHelper.parseAndFlattenJexlQuery(opt);
+            } catch (ParseException e) {
+                throw new RuntimeException(e);
+            }
+        }
+
+        if (options.containsKey(QueryOptions.DATATYPE_FILTER)) {
+            String opt = options.get(QueryOptions.DATATYPE_FILTER);
+            this.datatypeFilter = new HashSet<>(Splitter.on(',').splitToList(opt));
+        }
+
+        if (options.containsKey(QueryOptions.START_TIME) && options.containsKey(QueryOptions.END_TIME)) {
+            String opt = options.get(QueryOptions.START_TIME);
+            long start = Long.parseLong(opt);
+
+            opt = options.get(QueryOptions.END_TIME);
+            long stop = Long.parseLong(opt);
+
+            this.timeFilter = LongRange.of(start, stop);
+        } else {
+            throw new RuntimeException("date not set");
+        }
+
+        if (options.containsKey(QueryOptions.INDEXED_FIELDS)) {
+            String opt = options.get(QueryOptions.INDEXED_FIELDS);
+            this.indexedFields = new HashSet<>(Splitter.on(',').splitToList(opt));
+        } else {
+            throw new RuntimeException("indexed fields not set");
+        }
```

Comment:
maybe pull this code into a reusable place since its common to all the iterators?

Reply:
Noted as part of https://github.com/NationalSecurityAgency/datawave/issues/2916